pkg/applicationserver/io/packages/loradms/v1/package.go

```diff
@@ -44,57 +45,69 @@ func (p *DeviceManagementPackage) RegisterHandlers(s *runtime.ServeMux, conn *gr
 	errUplinkRequestFailed = errors.DefineInternal("uplink_request_failed", "uplink request failed")
+	errNoAssociation       = errors.DefineInternal("no_association", "no association available")
 )
 
 // HandleUp implements packages.ApplicationPackageHandler.
-func (p *DeviceManagementPackage) HandleUp(ctx context.Context, assoc *ttnpb.ApplicationPackageAssociation, up *ttnpb.ApplicationUp) error {
+func (p *DeviceManagementPackage) HandleUp(ctx context.Context, def *ttnpb.ApplicationPackageDefaultAssociation, assoc *ttnpb.ApplicationPackageAssociation, up *ttnpb.ApplicationUp) error {
 	ctx = log.NewContextWithField(ctx, "namespace", "applicationserver/io/packages/loradms/v1")
 	logger := log.FromContext(ctx)
 
+	if def == nil && assoc == nil {
+		return errNoAssociation.New()
+	}
+
 	if up.DevEUI == nil || up.DevEUI.IsZero() {
 		logger.Debug("Package configured for end device with no device EUI")
 		return nil
 	}
 
-	eui := objects.EUI(*up.DevEUI)
-	message := up.GetUplinkMessage()
-	settings := message.GetSettings()
-	uplink := objects.LoRaUplink{
-		FCnt:      message.GetFCnt(),
-		Port:      uint8(message.GetFPort()),
-		Payload:   objects.Hex(message.FRMPayload),
-		DR:        uint8(settings.DataRateIndex),
-		Freq:      uint32(settings.GetFrequency()),
-		Timestamp: float64(message.ReceivedAt.UTC().Unix()),
+	data, fPort, err := p.mergePackageData(def, assoc)
+	if err != nil {
+		return err
 	}
 
-	var data packageData
-	err := data.fromStruct(assoc.Data)
-	if errors.IsNotFound(err) {
-		// If there is no package data available, just reset the data.
-		data = packageData{}
-		assoc, err = p.registry.Set(ctx, assoc.ApplicationPackageAssociationIdentifiers, []string{"data"},
-			func(assoc *ttnpb.ApplicationPackageAssociation) (*ttnpb.ApplicationPackageAssociation, []string, error) {
-				assoc.Data = data.toStruct()
-				return assoc, []string{"data"}, nil
-			},
-		)
-		if err != nil {
-			logger.WithError(err).Debug("Failed to update package data")
-			return err
+	switch m := up.Up.(type) {
+	case *ttnpb.ApplicationUp_JoinAccept:
+		join := m.JoinAccept
+		loraUp := &objects.LoRaUplink{
+			Type:      objects.JoiningUplinkType,
+			Timestamp: toFloat64(float64(join.ReceivedAt.UTC().Unix())),
 		}
-	} else if err != nil {
-		logger.WithError(err).Debug("Failed to parse package data")
-		return err
+		return p.sendUplink(ctx, up, loraUp, data)
+	case *ttnpb.ApplicationUp_UplinkMessage:
+		msg := m.UplinkMessage
+		settings := msg.GetSettings()
+		loraUp := &objects.LoRaUplink{
+			Type:      objects.UplinkUplinkType,
+			FCnt:      toUint32(msg.GetFCnt()),
+			Port:      toUint8(uint8(msg.GetFPort())),
+			Payload:   objects.Hex(msg.FRMPayload),
+			DR:        toUint8(uint8(settings.DataRateIndex)),
+			Freq:      toUint32(uint32(settings.Frequency)),
+			Timestamp: toFloat64(float64(msg.ReceivedAt.UTC().Unix())),
+		}
+		if fPort != msg.FPort {
+			log.FromContext(ctx).Debug("Uplink received on unhandled FPort; drop payload")
```

Comment:
It's not dropping it; it's masking it
Also it's not really an unhandled port but rather a non-associated port (?)